### PR TITLE
ENH: Python parser now accepts delim_whitespace=True

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.

### CODE REVIEW

**Comment by jreback** [comment]
can you confirm with #12686 that all options are covered?

**Comment by jreback** [comment]
I think doc-string and io.rst need small edits for this (to remove that this is only for c engine)

**Comment by gfyoung** [comment]
I don't quite understand your question.  Could you clarify?

File: pandas/io/tests/test_parsers.py [hunk 1]
```diff
@@ -2867,6 +2841,103 @@ def test_read_only_header_no_rows(self):
         df = self.read_csv(StringIO('a,b,c'), index_col=False)
         tm.assert_frame_equal(df, expected)
 
+    def test_line_comment(self):
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: what is the net affect of moving all of these tests around?

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Better coverage.  Now all of the engines are getting tested.

**Comment by jreback** [comment]
I meant we are now listing all of the options in that issue so that we can check them off. And you have them listed in the variables to check for the options, just want a cross check.

bonus points if you can autogenerate a section in the doc-string which explicity lists the unsupported options (by using those variables)

**Comment by gfyoung** [comment]
Well, they used to be a copy of `_c_parser_defaults`, which I would consider to be the "correct" version until demonstrated otherwise.  So in fact the list in #12686 should be the one I created in this PR, which it is not.

**Comment by jreback** [comment]
@gfyoung if you can post a correct list I will update

**Comment by gfyoung** [comment]
FYI, Travis has just given the green light for my changes, so if you don't have any other code-related complaints, I'll update the documentation with `[ci skip]`.

**Comment by gfyoung** [comment]
<pre>'as_recarray',
'na_filter',
'compact_ints',
'use_unsigned',
'low_memory',
'memory_map',
'buffer_lines',
'error_bad_lines',
'warn_bad_lines',
'dtype',
'decimal',
'float_precision'<pre>

**Comment by jreback** [comment]
ok I edited the other issue, pls have a look

**Comment by gfyoung** [comment]
Are those uncapitalised headers (e.g. "undocumented in C engine only") sub sections for "Features supported in the C engine only"?  If not, all of the ones that I listed just now should also go under "Features supported in the C engine only"

File: doc/source/io.rst [hunk 2]
```diff
@@ -101,8 +101,7 @@ delim_whitespace : boolean, default False
   Specifies whether or not whitespace (e.g. ``' '`` or ``'\t'``)
   will be used as the delimiter. Equivalent to setting ``sep='\+s'``.
   If this option is set to True, nothing should be passed in for the
-  ``delimiter`` parameter. This parameter is currently supported for
-  the C parser only.
+  ``delimiter`` parameter.
```
Comment (jreback) [inline doc/source/io.rst]: add that this was added in 0.18.1

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -57,7 +57,7 @@
     Specifies whether or not whitespace (e.g. ``' '`` or ``'\t'``) will be
     used as the sep. Equivalent to setting ``sep='\+s'``. If this option
     is set to True, nothing should be passed in for the ``delimiter``
-    parameter. This parameter is currently supported for the C parser only.
+    parameter.
```
Comment (jreback) [inline pandas/io/parsers.py]: same here

File: pandas/io/parsers.py [hunk 4]
```diff
@@ -647,8 +660,13 @@ def _get_options_with_defaults(self, engine):
                 value = kwds[argname]
 
                 if engine != 'c' and value != default:
-                    raise ValueError('The %r option is not supported with the'
-                                     ' %r engine' % (argname, engine))
+                    if ('python' in engine and
```
Comment (jreback) [inline pandas/io/parsers.py]: ``engine == 'python'``

Comment (jreback) [inline pandas/io/parsers.py]: nvm. we have ``python-fwf`` as well

File: pandas/io/tests/test_parsers.py [hunk 5]
```diff
@@ -4652,6 +4655,9 @@ def test_invalid_c_parser_opts_with_not_c_parser(self):
         engines = 'python', 'python-fwf'
         for default in c_defaults:
             for engine in engines:
+                if 'python' in engine and default not in py_unsupported:
```
Comment (jreback) [inline pandas/io/tests/test_parsers.py]: does this fully test the converse. e.g. python only options on c-parser?

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: The test name indicates that the answer is no.  However, I do believe that the invalid parameters are tested but in different parts of the test suite (again, this is where a refactor would be good in a separate PR).

Comment (jreback) [inline pandas/io/tests/test_parsers.py]: ok another PR is fine.

ideally we have some tests which will explicity test for the unsupported features (so that when one IS defined in say the python-parser that was previously only in the c-parser) it will force an explict change (in the PR).

Comment (gfyoung) [inline pandas/io/tests/test_parsers.py]: Yep, I certainly agree.  Will make note for that PR.

File: pandas/io/parsers.py (hunk 4 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Yep, that's why I wrote it like that.

File: doc/source/io.rst (hunk 2 above)
Comment (jreback) [inline doc/source/io.rst]: @gfyoung if you can update this, and lmlk when you push.

Comment (gfyoung) [inline doc/source/io.rst]: Done.

File: pandas/io/parsers.py (hunk 3 above)
Comment (gfyoung) [inline pandas/io/parsers.py]: Done.

**Comment by jreback** [comment]
ty
